k2/python/csrc/torch/v2/ragged_arc.h

```diff
@@ -0,0 +1,93 @@
+/**
+ * @brief python wrapper for Ragged<Arc>
+ *
+ * @copyright
+ * Copyright      2021  Xiaomi Corp.  (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef K2_PYTHON_CSRC_TORCH_V2_RAGGED_ARC__H_
+#define K2_PYTHON_CSRC_TORCH_V2_RAGGED_ARC__H_
+
+#include <unordered_map>
+
+#include "k2/csrc/fsa.h"
+#include "k2/csrc/ragged.h"
+#include "k2/python/csrc/torch.h"
+#include "k2/python/csrc/torch/v2/ragged_any.h"
+
+namespace k2 {
+
+// It is a wrapper of Ragged<Arc> to support backward props in PyTorch
+struct RaggedArc {
+  Ragged<Arc> fsa;
+  torch::Tensor scores;  // shares the same memory with fsa.values
+
+  std::unordered_map<std::string, torch::Tensor> tensor_attrs;
```

Comment:
https://github.com/k2-fsa/k2/pull/814#discussion_r698141243

> but for me the main question in my mind is what is going to happen to class k2.Fsa. That has a bunch of
logic, e.g. about attributes, that I wonder how it will be transferred to C++; and once it is C++, who will be able to
maintain it!

I would propose to let these three `maps` manage all attributes of `k2.Fsa`.

We can add
- `__getattr__`
- `__setattr__`
to `RaggedArc` when wrapping it to Python. In this way, we can support assigning
arbitrary attributes to an Fsa in Python. 
- If the attribute is a tensor, we save it to `tensor_attrs`.
- If the attribute is a ragged tensor, we save it to `ragged_tensor_attrs`.
- Otherwise, we save it to `other_attrs`.

`all_attr_names` is to ensure that an attribute has exactly only one type.

Reply:
Ah, I see, I missed this.
I feel like the name RaggedArc is misleading, since this implements the full Fsa logic.
Maybe something like FsaWrapper or FsaClass would be clearer.
My main concern with this, I suppose, is that the code becomes less easy for others to understand,
view or maintain, because C++ is less accessible.  But it does solve a potential problem with deployment
in a C++-only environment.
Also, if we go down this road, it's going to be a lot of work, and I wonder whether it can be broken up
into stages somehow?  E.g. first adding the Any stuff, and later move class Fsa to C++?

Reply:
> Also, if we go down this road, it's going to be a lot of work, and I wonder whether it can be broken up
into stages somehow? E.g. first adding the Any stuff, and later move class Fsa to C++?

Yes, I agree.  I will try to get `RaggedAny` done first.

Reply:
> My main concern with this, I suppose, is that the code becomes less easy for others to understand,

I think this can be mitigated by detailed documentation.

Reply:
Sounds exciting!
@csukuangfj  You could submit `RaggedAny` first, I'll take part in some stuff of moving class Fsa to C++, if possible.

Reply:
@pkufool 
Thanks!